torchvision/csrc/cpu/image/readpng_cpu.cpp

```diff
@@ -72,30 +74,74 @@ torch::Tensor decodePNG(const torch::Tensor& data) {
     TORCH_CHECK(retval == 1, "Could read image metadata from content.")
   }
 
-  int channels;
-  switch (color_type) {
-    case PNG_COLOR_TYPE_RGB:
-      channels = 3;
-      break;
-    case PNG_COLOR_TYPE_RGB_ALPHA:
-      channels = 4;
-      break;
-    case PNG_COLOR_TYPE_GRAY:
-      channels = 1;
-      break;
-    case PNG_COLOR_TYPE_GRAY_ALPHA:
-      channels = 2;
-      break;
-    case PNG_COLOR_TYPE_PALETTE:
-      channels = 1;
-      break;
-    default:
-      png_destroy_read_struct(&png_ptr, &info_ptr, nullptr);
-      TORCH_CHECK(false, "Image color type is not supported.");
+  int current_channels = png_get_channels(png_ptr, info_ptr);
+
+  if (channels > 0 && channels != current_channels) {
+    // TODO: consider supporting PNG_INFO_tRNS
+    bool is_palette = (color_type & PNG_COLOR_MASK_PALETTE) != 0;
+    bool has_color = (color_type & PNG_COLOR_MASK_COLOR) != 0;
+    bool has_alpha = (color_type & PNG_COLOR_MASK_ALPHA) != 0;
+
+    switch (channels) {
+      case 1: // Gray or Palette
+        if (has_alpha) {
+          png_set_strip_alpha(png_ptr);
+        }
+
+        if (has_color && !is_palette) {
+          png_set_rgb_to_gray(png_ptr, 1, 0.2989, 0.587);
+        }
+        break;
+      case 2: // Gray + Alpha
+        if (is_palette) {
+          png_set_palette_to_rgb(png_ptr);
+          has_alpha = true;
+        }
+
+        if (!has_alpha) {
+          png_set_add_alpha(png_ptr, (1 << bit_depth) - 1, PNG_FILLER_AFTER);
+        }
+
+        if (has_color) {
+          png_set_rgb_to_gray(png_ptr, 1, 0.2989, 0.587);
+        }
+        break;
+      case 3:
+        if (is_palette) {
+          png_set_palette_to_rgb(png_ptr);
+          has_alpha = true;
+        } else if (!has_color) {
+          png_set_gray_to_rgb(png_ptr);
+        }
+
+        if (has_alpha) {
+          png_set_strip_alpha(png_ptr);
+        }
+        break;
+      case 4:
+        if (is_palette) {
+          png_set_palette_to_rgb(png_ptr);
+          has_alpha = true;
+        } else if (!has_color) {
+          png_set_gray_to_rgb(png_ptr);
+        }
+
+        if (!has_alpha) {
+          png_set_add_alpha(png_ptr, (1 << bit_depth) - 1, PNG_FILLER_AFTER);
+        }
+        break;
+      default:
+        png_destroy_read_struct(&png_ptr, &info_ptr, nullptr);
+        TORCH_CHECK(false, "Invalid number of output channels.");
+    }
+
+    png_read_update_info(png_ptr, info_ptr);
+  } else {
+    channels = current_channels;
   }
 
-  auto tensor = torch::empty(
-      {int64_t(height), int64_t(width), int64_t(channels)}, torch::kU8);
+  auto tensor =
+      torch::empty({int64_t(height), int64_t(width), channels}, torch::kU8);
```

Comment:
don't we need to keep the cast to avoid compiler warnings?

Reply:
No because channels is already an int64_t. No need to copy it.